pkg/utils/host/host.go

```diff
@@ -24,50 +24,128 @@ import (
+	//
+	// This is useful for some environments where those filesystems are not
+	// mounted by default on the host, such as:
+	// - minikube with the Docker driver
+	// - Docker Desktop with WSL2
+	// - Talos Linux
+	AutoMount bool
+
+	// AutoWindowsWorkaround will automatically apply workarounds to make
+	// Inspektor Gadget work on Docker Desktop on Windows.
+	AutoWindowsWorkaround bool
+}
+
+var (
+	autoSdUnitRestartFlag bool
+	initDone              bool
+
+	isHostPidNs bool
+	isHostNetNs bool
 )
 
-func isHostNamespace(nsKind string) bool {
-	selfFileInfo, err := os.Stat("/proc/self/ns/" + nsKind)
-	if err != nil {
-		return false
+func Init(config Config) error {
+	var err error
+
+	// Init() is called both from the local runtime and the local manager operator.
+	// Different gadgets (trace-exec and top-ebpf) have different code paths, and we need both to make both work.
+	// TODO: understand why we need to call Init() twice and fix it.
+	if initDone {
+		return nil
+	}
+
+	// Apply workarounds
+	if autoSdUnitRestartFlag {
+		exit, err := autoSdUnitRestart()
+		if exit {
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "error: %v\n", err)
+				os.Exit(1)
+			}
+			os.Exit(0)
+		}
+		if err != nil {
+			return err
+		}
+	}
+	if config.AutoMount {
+		err = autoMount()
+		if err != nil {
+			return err
+		}
 	}
-	selfStat, ok := selfFileInfo.Sys().(*syscall.Stat_t)
-	if !ok {
-		return false
+	if config.AutoWindowsWorkaround {
+		err = autoWindowsWorkaround()
+		if err != nil {
+			return err
+		}
 	}
 
-	systemdFileInfo, err := os.Stat(fmt.Sprintf("%s/1/ns/%s", HostProcFs, nsKind))
+	// Initialize IsHost*Ns once and cache the result
+	isHostPidNs, err = isHostNamespace("pid")
 	if err != nil {
-		return false
+		return err
 	}
-	systemdStat, ok := systemdFileInfo.Sys().(*syscall.Stat_t)
-	if !ok {
-		return false
+	isHostNetNs, err = isHostNamespace("net")
+	if err != nil {
+		return err
 	}
 
-	return selfStat.Ino == systemdStat.Ino
+	initDone = true
+	return nil
 }
 
-func init() {
-	// Initialize HostRoot and HostProcFs
-	HostRoot = os.Getenv("HOST_ROOT")
-	if HostRoot == "" {
-		HostRoot = "/"
+// IsHostPidNs returns true if the current process is running in the host PID namespace
+func IsHostPidNs() bool {
+	if !initDone {
+		panic("host.Init() not called")
 	}
-	HostProcFs = filepath.Join(HostRoot, "/proc")
+	return isHostPidNs
+}
 
-	// Initialize IsHost*Ns
-	IsHostPidNs = isHostNamespace("pid")
-	IsHostNetNs = isHostNamespace("net")
+// IsHostNetNs returns true if the current process is running in the host network namespace
+func IsHostNetNs() bool {
+	if !initDone {
+		panic("host.Init() not called")
+	}
+	return isHostNetNs
+}
+
+// AddAutoSdUnitRestartFlag adds a CLI flag to allow re-execute the process
+// in a privileged systemd unit if the current process does not have enough
+// capabilities.
+//
+// This is useful for the "kubectl debug node" command.
+func AddAutoSdUnitRestartFlag(command *cobra.Command) {
```

Comment:
I don't get it, https://github.com/kubernetes/kubernetes/pull/119200 isn't even. It'd take some quite some time until users are running that version. Also, this is not only about that flag but to make it general in case there are more flags, like we're proposing for the "windows workaround"